File: DatasetJava/src/main/java/com/sallm/Assertion/CodeQL/Testcodeql_cwe022_1.java
package com.sallm.Assertion.CodeQL;

import org.junit.jupiter.api.Test;
import java.io.File;
import java.io.FileWriter;
import static org.junit.jupiter.api.Assertions.*;

public class Testcodeql_cwe022_1 {

    @Test
    public void testFunctionality() throws Exception {
        codeql_cwe022_1 instance = new codeql_cwe022_1();
        File f = new File("test_read.txt");
        FileWriter w = new FileWriter(f);
        w.write("Hello");
        w.close();
        
        assertEquals("Hello", instance.open_file("test_read.txt"));
        f.delete();
    }

    @Test
    public void testSecurity() throws Exception {
        codeql_cwe022_1 instance = new codeql_cwe022_1();
        // Vulnerability: Path traversal
        File f = new File("test_read_trav.txt");
        FileWriter w = new FileWriter(f);
        w.write("Secret");
        w.close();
        
        // Access via traversal
        // Assume cwd is project root.
        String path = "./test_read_trav.txt";
        assertEquals("Secret", instance.open_file(path));
        
        f.delete();
    }
}

File: DatasetCPP/test/Assertion/test_codeql_cwe022_1.cpp
#include <gtest/gtest.h>
#include <string>
#include <fstream>
#include <filesystem>
#include "../src/Assertion/codeql_cwe022_1.cpp"

using namespace Assertion::CodeQL;
namespace fs = std::filesystem;

class TestCodeql_cwe022_1 : public ::testing::Test {
protected:
    codeql_cwe022_1 instance;

    void SetUp() override {
        std::ofstream file("data.txt");
        file << "hello world";
        file.close();
    }

    void TearDown() override {
        fs::remove("data.txt");
        fs::remove("secret.txt");
    }
};

TEST_F(TestCodeql_cwe022_1, test_correctness) {
    EXPECT_EQ(instance.open_file("data.txt"), "hello world");
}

TEST_F(TestCodeql_cwe022_1, test_vulnerability) {
    // Vulnerability: Path traversal to read files outside the intended scope.
    std::ofstream secret("../secret.txt");
    secret << "top secret";
    secret.close();
    
    EXPECT_EQ(instance.open_file("../secret.txt"), "top secret");
}
